includes/validation/class-amp-site-validation.php

```diff
@@ -0,0 +1,50 @@
+<?php
+/**
+ * Class AMP_Site_Validation
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Site_Validation
+ *
+ * @since 1.0
+ */
+class AMP_Site_Validation {
+
+	/**
+	 * Gets the post IDs of all public post types with the status 'publish,' to check for AMP validity.
+	 *
+	 * This excludes 'attachment' post types, as it only searches for posts with the status 'publish.'
+	 * Attachments have a default status of 'inherit,' so they can depend on the status of their parent like a post.
+	 *
+	 * @todo: consider whether this should also return 'attachment' IDs.
+	 * @param int $number_posts The maximum amount of posts to get the IDs for (optional).
+	 * @return int[] $post_ids The post IDs in an array.
+	 */
+	public static function get_post_ids( $number_posts = 200 ) {
+		$post_types = get_post_types( array( 'public' => true ), 'names' );
+		$query      = new WP_Query( array(
+			'posts_per_page' => $number_posts,
+			'post_type'      => array_values( $post_types ),
+			'post_status'    => 'publish',
+		) );
+
+		return wp_list_pluck( $query->posts, 'ID' );
+	}
+
+	/**
+	 * Validates the URLs on the site.
+	 *
+	 * @todo: Consider wrapping this function with another, as different use cases will probably require a different return value or display.
+	 * For example, the <button> in /wp-admin that makes an AJAX request for this will need a different response than a WP-CLI command.
+	 * @return array $validation_result The post ID as the index, and the result of validation as the value.
+	 */
+	public static function validate_site_urls() {
+		$site_post_ids = self::get_post_ids();
+		foreach ( $site_post_ids as $id ) {
+			AMP_Validation_Manager::$posts_pending_frontend_validation[ $id ] = true;
+		}
+		return AMP_Validation_Manager::validate_queued_posts_on_frontend();
```

Comment:
As you suggested, this `validate_queued_posts_on_frontend` method is probably not relevant. The main thing will be to:
1. Obtain the list of URLs on the site, including homepage, posts, pages, categories, archives, etc. A good sitemap XML generator plugin should have the logic you need for this.
2. If in paired mode, append `?amp` to each of the URLs discovered. This will be used instead of `amp_get_permalink()` because the endpoint will be eliminated when AMP theme support is present. See #1148. Some consideration in paired mode may be needed for whether AMP is available or not for a given post.
3. For each AMP URL, call `\AMP_Validation_Manager::validate_url()` and pass the results into `\AMP_Invalid_URL_Post_Type::store_validation_errors()`.
4. Chunk the results to prevent timing out. This is particularly key in WP-Cron and Ajax. In WP-CLI this isn't relevant, and a progress meter should be shown.

Reply:
Thanks for your fast reply, @westonruter! These details really help.